File: src/typescript/src/xiaoyi/core/config/source/file.ts
/**
 * # File Configuration Source
 *
 * `file` provides file-based configuration loading (JSON, YAML, TOML).
 *
 * Path: `xiaoyi.core.config.source.file`
 *
 * - Layer 0: `core`
 * - Layer 1: `config`
 * - Layer 2: `source`
 * - Layer 3: `file`
 *
 * @module core.config.source.file
 * @brief File-based configuration source
 * @group Core
 * @since 0.1.0
 * @author Miruamel
 * @see core.config.source
 * @see core.config.source.env
 * @see core.config.source.vault
 */
import { ConfigSource, ConfigSourceError } from ".";
import * as fs from "fs/promises";
import * as path from "path";

/** Supported config file formats. */
export enum ConfigFormat {
  /** JSON format. */
  Json = "json",
  /** YAML format. */
  Yaml = "yaml",
  /** TOML format. */
  Toml = "toml",
}

/** File source options. */
export interface FileSourceOptions {
  /** File path. */
  path: string;
  /** File format (auto-detected from extension if not specified). */
  format?: ConfigFormat;
  /** Watch for changes. */
  watch?: boolean;
  /** Source priority. */
  priority?: number;
}

/** File-based configuration source. */
export class FileSource implements ConfigSource {
  public readonly name: string;
  public readonly priority: number;
  private readonly filePath: string;
  private readonly format: ConfigFormat;
  private readonly watchEnabled: boolean;
  private watchers: Map<string, () => void> = new Map();

  /**
   * Create file source.
   *
   * @param options - Source options
   * @since 0.1.0
   */
  constructor(options: FileSourceOptions) {
    this.filePath = path.resolve(options.path);
    this.format = options.format ?? this.detectFormat();
    this.name = `file:${this.filePath}`;
    this.priority = options.priority ?? 100;
    this.watchEnabled = options.watch ?? false;
  }

  /** Detect format from extension. */
  private detectFormat(): ConfigFormat {
    const ext = path.extname(this.filePath).toLowerCase();
    switch (ext) {
      case ".json": return ConfigFormat.Json;
      case ".yaml":
      case ".yml": return ConfigFormat.Yaml;
      case ".toml": return ConfigFormat.Toml;
      default: return ConfigFormat.Json;
    }
  }

  /**
   * Load configuration from file.
   *
   * @returns Parsed configuration
   * @throws {ConfigSourceError} If loading or parsing fails
   * @since 0.1.0
   */
  async load(): Promise<Record<string, unknown>> {
    try {
      const content = await fs.readFile(this.filePath, "utf-8");
      return this.parse(content);
    } catch (error) {
      throw new ConfigSourceError(this.name, `Failed to load: ${error}`, error as Error);
    }
  }

  /** Parse content based on format. */
  private parse(content: string): Record<string, unknown> {
    switch (this.format) {
      case ConfigFormat.Json:
        return JSON.parse(content);
      case ConfigFormat.Yaml:
        // Simple YAML parsing - in production use a proper YAML library
        return this.parseYaml(content);
      case ConfigFormat.Toml:
        // Simple TOML parsing - in production use a proper TOML library
        return this.parseToml(content);
    }
  }

  /** Simple YAML parser (subset). */
  private parseYaml(content: string): Record<string, unknown> {
    const result: Record<string, unknown> = {};
    for (const line of content.split("\n")) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith("#")) continue;
      const idx = trimmed.indexOf(":");
      if (idx > 0) {
        const key = trimmed.slice(0, idx).trim();
        const value = trimmed.slice(idx + 1).trim();
        result[key] = this.parseYamlValue(value);
      }
    }
    return result;
  }

  /** Parse YAML value. */
  private parseYamlValue(value: string): unknown {
    if (value === "true") return true;
    if (value === "false") return false;
    if (value === "null" || value === "~") return null;
    const num = Number(value);
    if (!Number.isNaN(num)) return num;
    return value.replace(/^["']|["']$/g, "");
  }

  /** Simple TOML parser (subset). */
  private parseToml(content: string): Record<string, unknown> {
    const result: Record<string, unknown> = {};
    for (const line of content.split("\n")) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith("#")) continue;
      const idx = trimmed.indexOf("=");
      if (idx > 0) {
        const key = trimmed.slice(0, idx).trim();
        const value = trimmed.slice(idx + 1).trim();
        result[key] = this.parseYamlValue(value);
      }
    }
    return result;
  }

  /**
   * Watch for file changes.
   *
   * @param callback - Change callback
   * @returns Unsubscribe function
   * @since 0.1.0
   */
  watch(callback: (config: Record<string, unknown>) => void): () => void {
    if (!this.watchEnabled) {
      return () => {};
    }
    // In production, use fs.watch or chokidar
    const interval = setInterval(async () => {
      try {
        const config = await this.load();
        callback(config);
      } catch {
        // Ignore watch errors
      }
    }, 5000);

    return () => clearInterval(interval);
  }
}

File: src/python/xiaoyi/core/config/source/file.py
"""
# File Configuration Source

`file` provides file-based configuration loading (JSON, YAML, TOML).

Path: `xiaoyi.core.config.source.file`

- Layer 0: `core`
- Layer 1: `config`
- Layer 2: `source`
- Layer 3: `file`

@module xiaoyi.core.config.source.file
@brief File-based configuration source
@group Core
@since 0.1.0
@author Miruamel
@see xiaoyi.core.config.source
@see xiaoyi.core.config.source.env
@see xiaoyi.core.config.source.vault
"""

import json
import os
from pathlib import Path
from typing import Any, Dict, Optional
from enum import Enum

from ..config import ConfigSource, ConfigSourceError


class ConfigFormat(str, Enum):
    """Supported config file formats."""

    #: JSON format.
    JSON = "json"
    #: YAML format.
    YAML = "yaml"
    #: TOML format.
    TOML = "toml"


class FileSourceOptions:
    """
    File source options.

    @brief Options for file configuration source
    @group Core
    @since 0.1.0
    """

    def __init__(
        self,
        path: str,
        format: Optional[ConfigFormat] = None,
        watch: bool = False,
        priority: int = 100,
    ):
        """
        Create file source options.

        @param path File path
        @param format File format (auto-detected from extension if not specified)
        @param watch Watch for changes
        @param priority Source priority
        @since 0.1.0
        """
        self.path = path
        self.format = format
        self.watch = watch
        self.priority = priority


class FileSource(ConfigSource):
    """
    File-based configuration source.

    @brief File-based configuration source
    @group Core
    @since 0.1.0
    @example
    ```python
    source = FileSource(FileSourceOptions(path="./config.toml"))
    config = await source.load()
    ```
    """

    def __init__(self, options: FileSourceOptions):
        self._path = options.path
        self._format = options.format or self._detect_format()
        self._watch = options.watch
        self._priority = options.priority
        self._name = f"file:{self._path}"

    @property
    def name(self) -> str:
        return self._name

    @property
    def priority(self) -> int:
        return self._priority

    def _detect_format(self) -> ConfigFormat:
        ext = Path(self._path).suffix.lower()
        if ext == ".json":
            return ConfigFormat.JSON
        elif ext in (".yaml", ".yml"):
            return ConfigFormat.YAML
        elif ext == ".toml":
            return ConfigFormat.TOML
        return ConfigFormat.JSON

    async def load(self) -> Dict[str, Any]:
        """
        Load configuration from file.

        @return Configuration object
        @throws ConfigSourceError If loading fails
        @since 0.1.0
        """
        try:
            path = Path(self._path)
            if not path.exists():
                raise ConfigSourceError(self.name, f"File not found: {self._path}")

            content = path.read_text(encoding="utf-8")

            if self._format == ConfigFormat.JSON:
                return json.loads(content)
            elif self._format == ConfigFormat.YAML:
                try:
                    import yaml
                    return yaml.safe_load(content) or {}
                except ImportError:
                    raise ConfigSourceError(
                        self.name, "PyYAML required for YAML support"
                    )
            elif self._format == ConfigFormat.TOML:
                try:
                    import tomllib
                    return tomllib.loads(content)
                except ImportError:
                    try:
                        import tomli
                        return tomli.loads(content)
                    except ImportError:
                        raise ConfigSourceError(
                            self.name, "tomllib or tomli required for TOML support"
                        )
            return {}
        except ConfigSourceError:
            raise
        except Exception as e:
            raise ConfigSourceError(self.name, str(e), e)

    def watch(self, callback) -> None:
        """
        Watch for file changes (not implemented in async version).

        @param callback Change callback
        @returns No-op unsubscribe
        @since 0.1.0
        """
        # File watching would require async file system events
        # This is a placeholder for future implementation
        pass


__all__ = [
    "ConfigFormat",
    "FileSourceOptions",
    "FileSource",
]